contracts/liquidator/OnChainLiquidator.sol

```diff
@@ -0,0 +1,539 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+import "./vendor/@uniswap/v3-core/contracts/interfaces/callback/IUniswapV3FlashCallback.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/base/PeripheryPayments.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/base/PeripheryImmutableState.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/libraries/PoolAddress.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/libraries/CallbackValidation.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/libraries/TransferHelper.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/interfaces/ISwapRouter.sol";
+
+import "../CometInterface.sol";
+import "../ERC20.sol";
+import "../IWstETH.sol";
+
+interface IUniswapV2Router {
+    function swapExactTokensForTokens(
+        uint256 amountIn,
+        uint256 amountOutMin,
+        address[] calldata path,
+        address to,
+        uint256 deadline
+    ) external returns (uint256[] memory amounts);
+}
+
+// Balancer Interfaces
+interface IAsset {}
+
+interface IVault {
+    enum SwapKind { GIVEN_IN, GIVEN_OUT }
+
+    struct BatchSwapStep {
+        bytes32 poolId;
+        uint256 assetInIndex;
+        uint256 assetOutIndex;
+        uint256 amount;
+        bytes userData;
+    }
+
+    struct FundManagement {
+        address sender;
+        bool fromInternalBalance;
+        address payable recipient;
+        bool toInternalBalance;
+    }
+
+    function batchSwap(
+        SwapKind kind,
+        BatchSwapStep[] memory swaps,
+        IAsset[] memory assets,
+        FundManagement memory funds,
+        int256[] memory limits,
+        uint256 deadline
+    ) external payable returns (int256[] memory);
+}
+
+// Curve Interfaces
+interface ICurveRegistry {
+    function find_pool_for_coins(address _from, address _to) external returns (address);
+}
+
+interface IStableSwap {
+    function exchange(int128 i, int128 j, uint256 _dx, uint256 _min_dy) external payable returns (uint256);
+}
+
+/**
+ * @title XXX
+ * @notice XXX
+ * @author Compound
+ */
+contract OnChainLiquidator is IUniswapV3FlashCallback, PeripheryImmutableState, PeripheryPayments {
+    /** Errors */
+    error InsufficientAmountOut(address tokenIn, address tokenOut, uint256 amountIn, uint256 amountOut, uint256 amountOutMin, Exchange exchange, uint24 fee);
```

Comment:
- [ ] update to pass the poolConfig as an argument, instead of `exchange` and `fee`

Reply:
- [ ] same with Swap event